include/can.h

```diff
@@ -0,0 +1,380 @@
+	 */
+	u32_t id_type      : 1;
+	/** target state of the rtr bit */
+	u32_t rtr          : 1;
+	/** target state of the identifier */
+	union {
+		u32_t std_id   : 11;
+		u32_t ext_id   : 29;
+	};
+	/** rtr bit mask */
+	u32_t rtr_mask     : 1;
+	/** identifier mask*/
+	union {
+		u32_t std_id_mask  : 11;
+		u32_t ext_id_mask  : 29;
+	};
+} __packed;
+
+/**
+ * @typedef can_tx_callback_t
+ * @brief Define the application callback handler function signature
+ *
+ * @param error_flags status of the preformed send operation
+ */
+typedef void (*can_tx_callback_t)(u32_t error_flags);
+
+/**
+ * @typedef can_rx_callback_t
+ * @brief Define the application callback handler function signature
+ *        for receiving.
+ *
+ * @param received message
+ */
+typedef void (*can_rx_callback_t)(struct can_msg *msg);
+
+/**
+ * @brief Configure operation of a host controller.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param mode Operation mode
+ * @param bitrate bus-speed in Baud/s
+ *
+ * @retval 0 If successful.
+ * @retval -EIO General input / output error, failed to configure device.
+ */
+typedef int (*can_configure_t)(struct device *dev, enum can_mode mode,
+							   u32_t bitrate);
+
+/**
+ * @brief Perform data transfer to CAN bus.
+ *
+ * This routine provides a generic interface to perform data transfer
+ * to the can bus. Use can_write() for simple write.
+ * *
+ * @param dev          Pointer to the device structure for the driver instance.
+ * @param msg          Message to transfer.
+ * @param timeout      Waiting for empty tx mailbox timeout in ms or K_FOREVER.
+ * @param callback_isr Is called when message was sent or a transmission error
+ *                     occurred. If null, this function is blocking until
+ *                     message is sent.
+ *
+ * @retval 0 If successful.
+ * @retval CAN_TX_* on failure.
+ */
+typedef int (*can_send_t)(struct device *dev, struct can_msg *msg,
+						  s32_t timeout, can_tx_callback_t callback_isr);
+
+
+/**
+ * @brief Attach a message queue to a single or group of identifiers.
+ *
+ * This routine attaches a message queue to identifiers specified by
+ * a filter. Whenever the filter matches, the message is pushed to the queue
+ * If a message passes more than one filter the priotity of the match
+ * is hardware dependent.
+ * A message queue can be attached to more than one filter.
+ * The message queue must me initialized before.
+ * *
+ * @param dev    Pointer to the device structure for the driver instance.
+ * @param msgq   Pointer to the already initialized message queue.
+ * @param filter Pointer to a can_filter structure defining the id filtering.
+ *
+ * @retval filter id on success.
+ * @retval CAN_NO_FREE_FILTER if there is no filter left.
+ */
+typedef int (*can_attach_msgq_t)(struct device *dev, struct k_msgq *msg_q,
+								 struct can_filter *filter);
+
+/**
+ * @brief Attach a isr callback function to a single or group of identifiers.
```

Comment:
an isr (no ?)

Reply:
no I meant you wrote "a isr", but shouldn't it be "an isr"

The name is just fine actually.

Reply:
I created this function with further work on high level protocols in mind.
This reduces copying of the message and a high level protocol can easily attach himself to this specific identifiers.

Reply:
oh misspelling :-)